File: docs/vs-2015/snippets/csharp/VS_Snippets_VSSDK/dynamictoolboxmembers/cs/dynamictoolboxmemberspackage.cs
// VsPkg.cs : Implementation of ItemConfiguration
//

//<Snippet09>
using System;
using System.Diagnostics;
using System.Globalization;
using System.Runtime.InteropServices;
using System.ComponentModel.Design;
using Microsoft.Win32;
using Microsoft.VisualStudio.Shell.Interop;
using Microsoft.VisualStudio.OLE.Interop;
using Microsoft.VisualStudio.Shell;

//<Snippet06>
using System.Collections;
using System.ComponentModel;
using System.Drawing.Design;
using System.Reflection;
//</Snippet06>

namespace Company.ItemConfigurationCS
{
    //<Snippet07>
    // ...
    [PackageRegistration(UseManagedResourcesOnly = true)]
    // ...
    [DefaultRegistryRoot("Software\\Microsoft\\VisualStudio\\9.0")]
    // ...
    [InstalledProductRegistration(false, "#110", "#112", "1.0", IconResourceID = 400)]
    // ...
    [ProvideLoadKey("Standard", "1.0", "Package Name", "Company", 1)]
    // ...
    [ProvideMenuResource(1000, 1)]
    [Guid(GuidList.guidItemConfigurationPkgString)]
    [ProvideToolboxItems(1)]
    [ProvideToolboxItemConfiguration(typeof(ToolboxConfig))]
    public sealed class DynamicToolboxMembersPackage : Package
    //</Snippet07>
    {
        //<Snippet08>
        private ArrayList ToolboxItemList;
        private static readonly string CategoryTab =
            "ItemConfiguration Walkthrough CS";
        //</Snippet08>

        // ...
        //</Snippet09>
        
        // ...
        public DynamicToolboxMembersPackage()
        {
            Trace.WriteLine(string.Format(CultureInfo.CurrentCulture,
                "Entering constructor for: {0}", this.ToString()));

            //<Snippet11>
            this.ToolboxInitialized += new EventHandler(OnRefreshToolbox);
            this.ToolboxUpgraded += new EventHandler(OnRefreshToolbox);
            //</Snippet11>
        }

        //<Snippet10>
        // Add new instances of all ToolboxItems to the toolbox item list.
        void OnRefreshToolbox(object sender, EventArgs e)
        {
            IToolboxService service =
                GetService(typeof(IToolboxService)) as IToolboxService;
            IVsToolbox toolbox =
                GetService(typeof(IVsToolbox)) as IVsToolbox;

            // Remove target tab and all items under it.
            foreach (ToolboxItem item in service.GetToolboxItems(CategoryTab))
            {
                service.RemoveToolboxItem(item);
            }
            toolbox.RemoveTab(CategoryTab);

            // Recreate the target tab with the items from the current list. 
            foreach (ToolboxItem item in ToolboxItemList)
            {
                service.AddToolboxItem(item, CategoryTab);
            }
            service.SelectedCategory = CategoryTab;

            service.Refresh();
        }
        //</Snippet10>

        /////////////////////////////////////////////////////////////////////////////
        // Overriden Package Implementation

        // ...
        protected override void Initialize()
        {
            Trace.WriteLine (string.Format(CultureInfo.CurrentCulture,
                "Entering Initialize() of: {0}", this.ToString()));
            base.Initialize();

            // Add our command handlers for menu (commands must exist in the .vsct file)
            OleMenuCommandService mcs =
                GetService(typeof(IMenuCommandService)) as OleMenuCommandService;
            if ( null != mcs )
            {
                // Create the command for the menu item.
                CommandID menuCommandID = new CommandID(
                    GuidList.guidItemConfigurationCmdSet,
                    (int)PkgCmdIDList.cmdidMyCommand);
                MenuCommand menuItem = new MenuCommand(MenuItemCallback, menuCommandID );
                mcs.AddCommand( menuItem );
            }

            //<Snippet12>
            // Use the toolbox service to get a list of all toolbox items in
            // this assembly.
            ToolboxItemList = new ArrayList(
                ToolboxService.GetToolboxItems(this.GetType().Assembly, ""));
            if (null == ToolboxItemList)
            {
                throw new ApplicationException(
                    "Unable to generate a toolbox item listing for " +
                    this.GetType().FullName);
            }

            // Update the display name of each toolbox item in the list.
            Assembly thisAssembly = this.GetType().Assembly;
            foreach (ToolboxItem item in ToolboxItemList)
            {
                Type underlyingType = thisAssembly.GetType(item.TypeName);
                AttributeCollection attribs =
                    TypeDescriptor.GetAttributes(underlyingType);
                DisplayNameAttribute displayName =
                    attribs[typeof(DisplayNameAttribute)] as DisplayNameAttribute;
                if (displayName != null && !displayName.IsDefaultAttribute())
                {
                    item.DisplayName = displayName.DisplayName;
                }
            }
            //</Snippet12>
        }

        // ...
        //<Snippet13>
        private void MenuItemCallback(object sender, EventArgs e)
        {
            IVsPackage pkg = GetService(typeof(Package)) as IVsPackage;
            pkg.ResetDefaults((uint)__VSPKGRESETFLAGS.PKGRF_TOOLBOXITEMS);
        }
        //</Snippet13>
    }
}

File: docs/vs-2015/snippets/visualbasic/VS_Snippets_VSSDK/dynamictoolboxmembers/vb/dynamictoolboxmemberspackage.vb
' Implementation of ItemConfiguration
'

'<Snippet09>
Imports Microsoft.VisualBasic
Imports System
Imports System.Diagnostics
Imports System.Globalization
Imports System.Runtime.InteropServices
Imports System.ComponentModel.Design
Imports Microsoft.Win32
Imports Microsoft.VisualStudio.Shell.Interop
Imports Microsoft.VisualStudio.OLE.Interop
Imports Microsoft.VisualStudio.Shell

'<Snippet06>
Imports System.Collections
Imports System.ComponentModel
Imports System.Drawing.Design
Imports System.Reflection
'</Snippet06>

'<Snippet07>
' ...
<PackageRegistration(UseManagedResourcesOnly:=True), _
    DefaultRegistryRoot("Software\\Microsoft\\VisualStudio\\9.0"), _
    InstalledProductRegistration(False, "#110", "#112", "1.0", IconResourceID:=400), _
    ProvideLoadKey("Standard", "1.0", "Package Name", "Company", 1), _
    ProvideMenuResource(1000, 1), _
    Guid(GuidList.guidItemConfigurationPkgString)> _
    <ProvideToolboxItems(1)> _
    <ProvideToolboxItemConfiguration(GetType(ToolboxConfig))> _
Public NotInheritable Class DynamicToolboxMembersPackage
    Inherits Package
    '</Snippet07>

    '<Snippet08>
    Private ToolboxItemList As ArrayList
    Private Shared ReadOnly CategoryTab As String = _
        "ItemConfiguration Walkthrough VB"
    '</Snippet08>

    ' ...
    '</Snippet09>

    ' ...
    Public Sub New()
        Trace.WriteLine(String.Format(CultureInfo.CurrentCulture, _
            "Entering constructor for: {0}", Me.GetType().Name))
    End Sub


    '<Snippet10>
    ' Add new instances of all ToolboxItems to the toolbox item list.
    Private Sub OnRefreshToolbox(ByVal sender As Object, ByVal e As EventArgs) _
    Handles Me.ToolboxInitialized, Me.ToolboxUpgraded

        Dim service As IToolboxService = _
            TryCast(GetService(GetType(IToolboxService)), IToolboxService)
        Dim toolbox As IVsToolbox = _
             TryCast(GetService(GetType(IVsToolbox)), IVsToolbox)

        ' Remove target tab and all items under it.
        For Each item As ToolboxItem In service.GetToolboxItems(CategoryTab)
            service.RemoveToolboxItem(item)
        Next
        toolbox.RemoveTab(CategoryTab)

        ' Recreate the target tab with the items from the current list. 
        For Each item As ToolboxItem In ToolboxItemList
            service.AddToolboxItem(item, CategoryTab)
        Next
        service.SelectedCategory = CategoryTab

        service.Refresh()
    End Sub
    '</Snippet10>


    ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    ' Overriden Package Implementation

    ''' <summary>
    ''' Initialization of the package; this method is called right after the package is sited, so this is the place
    ''' where you can put all the initilaization code that rely on services provided by VisualStudio.
    ''' </summary>
    Protected Overrides Sub Initialize()
        Trace.WriteLine(String.Format(CultureInfo.CurrentCulture, "Entering Initialize() of: {0}", Me.GetType().Name))
        MyBase.Initialize()

        ' Add our command handlers for menu (commands must exist in the .vsct file)
        Dim mcs As OleMenuCommandService = TryCast(GetService(GetType(IMenuCommandService)), OleMenuCommandService)
        If Not mcs Is Nothing Then
            ' Create the command for the menu item.
            Dim menuCommandID As New CommandID(GuidList.guidItemConfigurationCmdSet, CInt(PkgCmdIDList.cmdidMyCommand))
            Dim menuItem As New MenuCommand(New EventHandler(AddressOf MenuItemCallback), menuCommandID)
            mcs.AddCommand(menuItem)
        End If

        '<Snippet12>
        ' Use the toolbox service to get a list of all toolbox items in
        ' this assembly.
        ToolboxItemList = New ArrayList( _
            ToolboxService.GetToolboxItems(Me.GetType().Assembly, ""))
        If ToolboxItemList Is Nothing Then
            Throw New ApplicationException( _
                "Unable to generate a toolbox item listing for " & _
                Me.GetType().FullName)
        End If

        ' Update the display name of each toolbox item in the list.
        Dim thisAssembly As Assembly = Me.GetType().Assembly
        For Each item As ToolboxItem In ToolboxItemList

            Dim underlyingType As Type = thisAssembly.GetType(item.TypeName)
            Dim attribs As AttributeCollection = _
                TypeDescriptor.GetAttributes(underlyingType)
            Dim displayName As DisplayNameAttribute = _
                TryCast(attribs(GetType(DisplayNameAttribute)), DisplayNameAttribute)

            If displayName IsNot Nothing AndAlso Not displayName.IsDefaultAttribute() Then
                item.DisplayName = displayName.DisplayName
            End If
        Next
        '</Snippet12>
    End Sub

    ''' <summary>
    ''' This function is the callback used to execute a command when the a menu item is clicked.
    ''' See the Initialize method to see how the menu item is associated to this function using
    ''' the OleMenuCommandService service and the MenuCommand class.
    ''' </summary>
    '<Snippet13>
    Private Sub MenuItemCallback(ByVal sender As Object, ByVal e As EventArgs)
        Dim pkg As IVsPackage = TryCast(GetService(GetType(Package)), IVsPackage)
        pkg.ResetDefaults(CUInt(__VSPKGRESETFLAGS.PKGRF_TOOLBOXITEMS))
    End Sub
    '</Snippet13>

End Class